Initial identity, shim, evaluation logic, reverse proxy, and key-vendor implementations

gkelly opened the pull request:
The final commit in the chain isn't ready for commit yet, it's mostly untested while we're getting our Elixir credentials in order.

gkelly committed:
identity: Initial definition

gkelly committed:
context: Initial implementation

gkelly committed:
validator: Interface and implementations

gkelly committed:
shim: Elixir and Static implementations

gkelly committed:
parser: Initial multi-issuer implementation

gkelly committed:
gcp/warehouse: Initial warehouse implementation

There are several unresolved issues with this implementation, such as
role management and key rotation.

gkelly committed:
Add Travis configuration

gkelly committed:
evaluator: Initial implementation

Evaluator combines a Parser and Validator into a single operation.

gkelly committed:
builder: Add builder.Build

Allow callers to programmatically construct a Validator from a protocol
buffer defintion of one.

kemp-google commented on builder/build.go:
```diff
@@ -0,0 +1,116 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package builder
```
builder.Build stutters
Is there a better name?

Also missing package comment

gkelly replied:
I agree it stutters, but I don't have a better name.  It feels like Reader.Read or Writer.Write.

Fixed missing package comment.

pgrosu replied:
either `getEvaluator` or `buildEvaluator` would be fine.

kemp-google commented on builder/build.go:
```diff
@@ -0,0 +1,116 @@
+
+// Build constructs a ga4gh.Evaluator from the protocol buffer definition of an
+// Evaluator.
+func Build(ctx context.Context, c *Evaluator) (*ga4gh.Evaluator, error) {
```
Why is an Evaluator a c?

gkelly replied:
Renamed to 'e'.

kemp-google commented on builder/builder.pb.go:
```diff
@@ -0,0 +1,611 @@
+// Copyright 2018 Google LLC
```
Is checking in the generated code actually the recommended Thing To Do?

gkelly replied:
Yes, other packages do the same.

kemp-google commented on context.go:
```diff
@@ -0,0 +1,33 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ga4gh
```
I wonder if you are better to have a sub package called ga4gh instead of this odd renaming

or even just identity

gkelly replied:
The renaming isn't actually renaming, it's go-vim acting up on me.  I'll remove the explicit renames.

kemp-google replied:
As discussed offline, we definitely don't want to hide the renaming so please put the names back.  It would be best if it wasn't surprising but I understand you don't want another ga4gh level in the tree.

pgrosu replied:
See my [previous suggestion](https://github.com/googlegenomics/ga4gh-identity/pull/1#discussion_r217899837) of either `getEvaluator` or `buildEvaluator`, either of which would be fine.

gkelly replied:
Re-added alias names.

kemp-google commented on context.go:
```diff
@@ -0,0 +1,33 @@
+const identityKey key = 0
+
+// NewContext creates a new context.Conext from ctx that carries identity.
+func NewContext(ctx context.Context, identity *Identity) context.Context {
```
I think these are usually called WithIdentity or NewIdentityContext or similar

gkelly replied:
I was following the examples from https://blog.golang.org/context and https://golang.org/pkg/context/.  gRPC uses the same style: https://godoc.org/google.golang.org/grpc/peer, and x/net/trace: https://godoc.org/golang.org/x/net/trace.

kemp-google replied:
It's true but all of those have useful package names: userip.NewContext or trace.NewContext etc - you have ga4gh.NewContext (but what's a ga4gh context?)

gkelly replied:
I'm convinced.  Renamed to NewIdentityContext/IdentityFromContext.

kemp-google commented on doc.go:
```diff
@@ -0,0 +1,18 @@
+// Copyright 2018 Google LLC
```
Why split things into so many files off the get go?  Each file is pretty small

gkelly replied:
I envisioned this having a lot more content, and then didn't write it.  Merging into identity.go.

kemp-google commented on evaluator.go:
```diff
@@ -0,0 +1,51 @@
+
+// Evaluator combines both parsing and validation of authorization tokens.
+type Evaluator struct {
+	Parser    *Parser
```
Do you explicitly not want embedding?  Otherwise drop the names.

gkelly replied:
I explicitly do not want embedding.

kemp-google commented on gcp/proxy/main.go:
```diff
@@ -0,0 +1,194 @@
+	certPath = flag.String("cert", "cert.pem", "TLS serving certificate")
+	keyPath  = flag.String("key", "key.pem", "TLS serving key")
+
+	proxyHost = flag.String("proxy_host", ":8080", "port to serve HTTPS proxy on")
```
Optional, but flags outside Google are almost always spelled-like-this

gkelly replied:
Acknowledged.  Dropping this proxy from this review, will update.

kemp-google commented on gcp/proxy/main.go:
```diff
@@ -0,0 +1,194 @@
+	}
+
+	// TODO: Reduce the scope required to operate the proxy.
+	client, err := google.DefaultClient(oauth2.NoContext, "https://www.googleapis.com/auth/cloud-platform")
```
Why no context?

gkelly replied:
It's calling from main(), so I could pass context.Background() or oauth2.NoContext.  NoContext seemed more semantically correct.  Do you think I should change it?

kemp-google commented on gcp/warehouse.go:
```diff
@@ -0,0 +1,158 @@
+	Scopes      []string
+}
+
+// Warehouse is used to create Google Cloud Platform Service Account keys and
```
gcp.Warehouse doesn't tell me much as a reader.
gcp.AccountManager might, or gcp.AccountWarehouse? 

even, though maybe overkill, gcp/accounts.Warehouse

gkelly replied:
Renamed to gcp.AccountWarehouse.

kemp-google commented on gcp/warehouse.go:
```diff
@@ -0,0 +1,158 @@
+		return nil, fmt.Errorf("getting backing account: %v", err)
+	}
+
+	// TODO: Remove old keys from service account before creating new keys.
```
Use TODO(#X) style and file issues

What happens as it stands?  Do the old keys get disabled automatically or new key creation fails?

gkelly replied:
New key creation fails.  Filed an issue, changed TODO.

kemp-google commented on gcp/warehouse.go:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	// TODO: Remove old keys from service account before creating new keys.
+	svc := wh.iam.Projects.ServiceAccounts.Keys
```
why svc instead of 'keys'?

result, err := keys.Create...

gkelly replied:
Keys is a ProjectsServiceAccountsKeyService, which is where that came from. Renamed to keys.

kemp-google commented on shim/doc.go:
```diff
@@ -0,0 +1,15 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shim
```
Missing documentation?

gkelly replied:
Done.

pgrosu replied:
Still not seeing the file :(

kemp-google commented on shim/elixir.go:
```diff
@@ -0,0 +1,79 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shim
```
I question having a number of shims in a single package.  What about something like shims/elixir

gkelly replied:
Done.

kemp-google commented on shim/injector.go:
```diff
@@ -0,0 +1,31 @@
+)
+
+// Injector is a ga4gh.Shim that returns a single fixed Identity.
+type Injector struct {
```
Injector seems an odd name to me.  Why not say what it does - eg Fixed/Single/Static?

gkelly replied:
I like Static.  Done.

kemp-google reviewed (commented):
Some initial commentary - let me know if a VC would be useful

gkelly commented:
I dropped the final commit in the chain.  I intend to submit this PR once reviews are complete now.  The proxy is now a reverse proxy will be in a subsequent review.

pgrosu commented on validator/boolean.go:
```diff
@@ -0,0 +1,68 @@
+// returns true.
+type Or []ga4gh.Validator
+
+// Validate returns true iff one of the validators that this Or wraps returns
```
This is not an English sentence for most users.  I know what you meant to say, but there are other ways to write it with easier clarity.

pgrosu commented on validator/boolean.go:
```diff
@@ -0,0 +1,68 @@
+	return false, nil
+}
+
+// And is a ga4gh.Validator that succeeds iff all of the wrapped validators do
```
Try to rewrite this for easier clarity for folks to understand (same as above regarding `Or`).

pgrosu commented on validator/boolean_test.go:
```diff
@@ -0,0 +1,140 @@
+		},
+		OriginOrganization: []ga4gh.StringValue{
+			{
+				Value: "Earth",
```
You have a good sense of humor :)

pgrosu commented on validator/simple.go:
```diff
@@ -0,0 +1,54 @@
+
+// Validate returns true iff there is a corresponding field in the input that
+// contains at least one value that matches each of the key and values stored
+// in the Simple.
```
You can drop `the`  (i.e. `s/in the Simple/in Simple`) since `Simple` becomes the noun.

gkelly committed:
gcp/proxy: Initial implementation

Initial implementation of a reverse proxy that rewrites Authorization
headers containing bearer tokens into Google Cloud Platform access
tokens.

kemp-google commented on builder/build.go:
```diff
@@ -0,0 +1,119 @@
+
+// Build constructs a ga4gh.Evaluator from the protocol buffer definition of an
+// Evaluator.
+func Build(ctx context.Context, e *Evaluator) (*ga4gh.Evaluator, error) {
```
I still feel like this package is badly named.  Maybe it's evaluator.Build?

gkelly replied:
How do you feel about evaluator.Build and then moving the proto definition into a evaluator/proto?  Just to prevent having types evaluator.Evaluator, evaluator.Shim, evaluator.Parser, etc.

kemp-google commented on builder/build.go:
```diff
@@ -0,0 +1,119 @@
+func buildParser(ctx context.Context, p *Parser) (*ga4gh.Parser, error) {
+	var shims []ga4gh.Shim
+	for _, shim := range p.Shims {
+		gs, err := buildShim(ctx, shim)
```
What's a gs?

pgrosu replied:
That stands for a `ga4gh.Shim`

gkelly replied:
ga4gh.Shim

kemp-google commented on builder/build.go:
```diff
@@ -0,0 +1,119 @@
+
+	switch v := v.Validator.(type) {
+	case *Validator_And_:
+		vs, err := buildValidators(ctx, v.And.Validators)
```
What's a vs?  I'm just wondering what these mnemonics are supposed to be (otherwise just use 'v' or 'result', etc)

pgrosu replied:
validators (plural)

gkelly replied:
Validators.

kemp-google commented on builder/build.go:
```diff
@@ -0,0 +1,119 @@
+	for _, v := range vs {
+		gv, err := buildValidator(ctx, v)
+		if err != nil {
+			return nil, fmt.Errorf("building validators: %v", err)
```
Include the validator that failed to build in the error?

gkelly replied:
Done.

kemp-google commented on builder/build_test.go:
```diff
@@ -0,0 +1,170 @@
+)
+
+var id = &ga4gh.Identity{
+	OriginOrganization: []ga4gh.StringValue{
```
Cuddle this up a bit more? { Value: "Earth" } etc

gkelly replied:
Done.

kemp-google commented on builder/builder.proto:
```diff
@@ -0,0 +1,55 @@
+
+syntax = "proto3";
+
+package builder;
```
This definitely feels like the wrong name.  These are DURI types aren't they?  Or identity types?  Or ga4gh types at least?

gkelly replied:
They're proto-duplicates of the toplevel types from the ga4gh package.

kemp-google commented on evaluator.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+// Evaluator combines both parsing and validation of authorization tokens.
+type Evaluator struct {
```
Should this be called IdentityEvaluator?

gkelly replied:
I think Evaluator fits with the other other names in the package, like Shim, Parser, etc.  They're all identity-related, so I don't think they need the prefix.

kemp-google replied:
OK, I still think this package is poorly named and we really want something like identity.Evaluator, etc.  From a user point of view, you'll have things like ga4gh.Evaluator and ga4gh.Shim which lack the important context in their name.  I'd be less picky about it except this is going to end up as a public API.

However, I'll leave it up to the maintainers to decide what to do about it since it won't be you or I.

kemp-google reviewed (commented):
Sorry I still haven't made it all the way through, I'll look again in the morning

kemp-google commented on builder/builder.proto:
```diff
@@ -0,0 +1,55 @@
+// Copyright 2018 Google LLC
```
Are these types derived directly from a GA4GH spec or no?  If they are, please link to the spec.  If they aren't please document.

gkelly replied:
Added package comment.

kemp-google commented on gcp/warehouse.go:
```diff
@@ -0,0 +1,160 @@
+		return "", fmt.Errorf("getting backing account: %v", err)
+	}
+
+	response, err := wh.creds.Projects.ServiceAccounts.GenerateAccessToken(accountID("-", account), &iamcredentials.GenerateAccessTokenRequest{
```
Why "-" in the call to accountID in a couple places?

pgrosu replied:
@kemp-google Because that's the definition as noted at the following line:

https://github.com/googleapis/google-api-go-client/blob/master/iamcredentials/v1/iamcredentials-gen.go#L114

pgrosu replied:
Or in the doc itself:  

https://godoc.org/google.golang.org/api/iamcredentials/v1#GenerateAccessTokenRequest

gkelly replied:
"-" is allowed when IAM can infer the project from the service account ID.

kemp-google commented on gcp/warehouse.go:
```diff
@@ -0,0 +1,160 @@
+	if err == nil {
+		return account.Email, nil
+	}
+	gErr, ok := err.(*googleapi.Error)
```
why not
if err, ok := err.(*googleapi.Error); !ok || err.Code != http.StatusNotFound {
 ...
}

gkelly replied:
Done.

kemp-google commented on gcp/warehouse.go:
```diff
@@ -0,0 +1,160 @@
+	return account.Email, nil
+}
+
+func hashID(id string) string {
```
Is this hash specified for use by GCP somewhere?

pgrosu replied:
@kemp-google He's doing something more simple than you imagine.  All Garret doing is randomizing the string so that it's not guessed.  He's using the get call to check if it exists, and if not then just create it.  Take a look at the following page for more information about service accounts:

https://cloud.google.com/iam/docs/service-accounts

Specifically this:


`SERVICE_ACCOUNT_NAME@PROJECT_ID.iam.gserviceaccount.com`

_You can create up to 100 service accounts per project (including the default Compute Engine service account and the App Engine service account) using the IAM API, the GCP console, or the gcloud command-line tool. These default service accounts and the service accounts you explicitly create are the user-managed service accounts._

gkelly replied:
No, this is not GCP specified.  I'm creating a deterministic mapping between service account IDs (which must match /^[a-z][a-z0-9]{,29}$/) and 'sub' claims.  The values aren't secret, they just need to not collide fairly well.

kemp-google commented on handler.go:
```diff
@@ -0,0 +1,54 @@
+
+// ServeHTTP implements the http.Handler interface.
+func (h *Handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	parts := strings.SplitN(req.Header.Get("authorization"), " ", 2)
```
Get() is case insensitive but the returned value isn't converted to lower case I don't think?

gkelly replied:
Good catch. Done.

kemp-google commented on handler.go:
```diff
@@ -0,0 +1,54 @@
+func (h *Handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	parts := strings.SplitN(req.Header.Get("authorization"), " ", 2)
+	if len(parts) != 2 || parts[0] != "bearer" {
+		log.Printf("not bearer token")
```
Cruft?

gkelly replied:
Oops. Done.

kemp-google commented on handler.go:
```diff
@@ -0,0 +1,54 @@
+	ctx := req.Context()
+	id, err := h.Evaluator.Evaluate(ctx, parts[1])
+	if err != nil {
+		log.Printf("evaluator failed: %v", err)
```
Cruft?

pgrosu replied:
Probably just practical to the scope of error-space

gkelly replied:
Also oops.  Done.

kemp-google commented on parser.go:
```diff
@@ -0,0 +1,113 @@
+		return nil, fmt.Errorf("parsing JWT: %v", err)
+	}
+
+	// This might look strange.  We're extracting the claims from the provided
```
This comment isn't very helpful to me.  It does look strange, but this doesn't tell me *why* you are ignoring the signature or expiry values.  Aren't they still meaningful even if you don't yet know what claim you want to use?

gkelly replied:
Yes, we could do manual verification of some things here.  But having the full IDTokenVerifier, which we can get from the 'iss' claim, will do that for us.

Updated comment.

kemp-google commented on parser.go:
```diff
@@ -0,0 +1,113 @@
+// Parse takes an authorization string (usually an HTTP authorization bearer
+// token) and converts it into an Identity.
+func (p *Parser) Parse(ctx context.Context, auth string) (*Identity, error) {
+	if err := p.lazyInit(ctx); err != nil {
```
Use sync.Once?

Also, it's a bit odd to arbitrarily pass the first context as the initialization context.  It's probably OK, but it's odd.

pgrosu replied:
Why?

gkelly replied:
Refactored to have NewParser and use a context there.

kemp-google commented on parser.go:
```diff
@@ -0,0 +1,113 @@
+	for issuer, clientID := range p.Issuers {
+		provider, err := oidc.NewProvider(ctx, issuer)
+		if err != nil {
+			return fmt.Errorf("creating issuer for %q: %v", issuer, err)
```
creating provider?

gkelly replied:
Done.

kemp-google commented on validator/boolean.go:
```diff
@@ -0,0 +1,68 @@
+	for _, v := range or {
+		ok, err := v.Validate(ctx, identity)
+		if err != nil {
+			return false, fmt.Errorf("nested validator: %v", err)
```
I think it will be hard to track down problems unless you include some indication of what validator failed

gkelly replied:
Added index to the error string.

kemp-google commented on validator/boolean.go:
```diff
@@ -0,0 +1,68 @@
+		if err != nil {
+			return false, fmt.Errorf("nested validator: %v", err)
+		}
+		results = append(results, ok)
```
result |= ok instead of the list and loop?  (same for and?)

gkelly replied:
Changed the wording to allow early returns and just did that.

kemp-google commented on validator/boolean_test.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+func ExampleOr() {
+	id := &ga4gh.Identity{
```
Optional, but for these examples I think they'll be more readable cuddled up a bit more

id := &ga4gh.Identity{Role: []ga4gh.StringValue{{Value: "human"}}}
v := Or{
  &Simple{"Role": "human"},
  &Simple{"Role": "robot"},
}

etc

gkelly replied:
Done.

kemp-google commented on validator/constant.go:
```diff
@@ -0,0 +1,32 @@
+
+// Constant is a ga4gh.Validator that returns a set success and error value.
+type Constant struct {
+	Ok  bool
```
I think 'OK' in Go

gkelly replied:
Done.

kemp-google reviewed (commented):
Still didn't make it all the way through, but more comments.

gkelly committed:
gcp: Make AccountWarehouse assign roles (#3)

The logic here allows assigning roles in other projects.  This works so
long as the identity running the service has IAM Admin rights in the
other project.

gkelly committed:
gcp: Refactor out common configuration logic

Add a new internal package 'appengine' for constructing common
configurable elements, like evaluators and warehouses.

gkelly committed:
gcp/key-vendor: Initial implementation

Add a key-vendor service which dispenses service account keys to callers
in exchange for their GA4GH identities.

gkelly commented:
Merged other PRs into this one to cut down on rebasing.

kemp-google commented on gcp/key-vendor/app.yaml:
```diff
@@ -0,0 +1,22 @@
+  # PROJECT determines which project backing service accounts for incoming
+  # requests are created in.
+  PROJECT: "your-gcp-project-here"
+  # ROLE is assigned to backing accounts as they are created.  The provided
```
s/backing/service ?

gkelly replied:
Added 'service'.

kemp-google commented on gcp/key-vendor/app.yaml:
```diff
@@ -0,0 +1,22 @@
+      }
+    }
+    validator {
+      simple {}
```
Why not something like constant { true }?  I realize it's the same thing but it's not clear unless you read the source for the simple validator

gkelly replied:
You're right, it's unclear.  I've added 'constant' to the builder proto.  I also toyed with the idea of making an empty Validator proto mean always succeeds, but don't like the idea of failing open.

kemp-google commented on gcp/internal/appengine/config.go:
```diff
@@ -0,0 +1,76 @@
+func mustGetenv(key string) string {
+	v := os.Getenv(key)
+	if v == "" {
+		log.Fatalf("Environment variable %q must be set, see app.yaml for more information.", key)
```
s/,/:/ and drop period?

gkelly replied:
Done.

kemp-google commented on builder/builder.proto:
```diff
@@ -0,0 +1,58 @@
+
+syntax = "proto3";
+
+// Package builder provides protocol buffer versions of some of the top-level
```
GA4GH has largely standardized on JSON for interchange.  Should we just drop the protos and add JSON tags to the in memory types?

gkelly replied:
I agree on JSON for interchange, but this is for local configuration and never on the API boundary.  To rebuild the oneof usage in JSON would require writing string matching and selecting which struct to JSON-unmarshal on top of, so I think it's much cleaner to keep this as proto.

kemp-google commented on validator/simple_test.go:
```diff
@@ -0,0 +1,90 @@
+	ga4gh "github.com/googlegenomics/ga4gh-identity"
+)
+
+func TestSimpleTrue(t *testing.T) {
```
SimpleTrue is an odd name considering it has test cases that return false.  Just Simple?

gkelly replied:
Done.

kemp-google commented on validator/simple.go:
```diff
@@ -0,0 +1,53 @@
+// identity to match the keys and values it contains.  For example, the Simple
+// validator: &Simple{"Role": "human"} would validate all identities containing
+// at least one Role with the value "human".  The Simple Validator does not
+// inspect the source of claims.
```
Is it worth noting here that the values are not plain values but must be GA4GH Values?  Wasn't clear to me until I read the source for the Validate function.  Otherwise this will quietly fail if you provide a map of plain values.

gkelly replied:
It only works if you pass it plain values, the Validate function takes care of unwrapping the matched values from their ga4gh.BoolValue/StringValue.

kemp-google commented on validator/simple.go:
```diff
@@ -0,0 +1,53 @@
+		}
+		var matched bool
+		for i := 0; i < field.Len(); i++ {
+			if reflect.DeepEqual(field.Index(i).FieldByName("Value").Interface(), expected) {
```
This code is panicy I think: FieldByName says it panics if the value isn't a struct.
I think Interface() also panics if FieldByName("Value") returns the zero value.

gkelly replied:
Yes, if you match non-ga4gh.{Bool,String}Value struct members this panics.  Added logic so that it only tries to reach into *Values if it knows about them.  Added test and updated example.
